crates/fuzzing/src/oracles.rs

```diff
@@ -726,13 +726,150 @@ pub fn differential_wasmi_execution(wasm: &[u8], config: &crate::generators::Con
+    // interfere, observable as an uncaught `SIGSEGV`--not even caught by
+    // libFuzzer. By running Wasmtime second, its signal handlers are registered
+    // most recently and they catch failures appropriately.
+    let spec_vals = wasm_spec_interpreter::interpret(wasm, vec![]);
+    debug!("spec interpreter returned: {:?}", &spec_vals);
+    let wasmtime_vals = run_in_wasmtime(wasm, config, &[]);
+    debug!("Wasmtime returned: {:?}", wasmtime_vals);
+
+    // Match a spec interpreter value against a Wasmtime value. Eventually this
+    // should support references and `v128` (TODO).
+    fn matches(spec_val: &wasm_spec_interpreter::Value, wasmtime_val: &wasmtime::Val) -> bool {
+        match (spec_val, wasmtime_val) {
+            (wasm_spec_interpreter::Value::I32(a), wasmtime::Val::I32(b)) => a == b,
+            (wasm_spec_interpreter::Value::I64(a), wasmtime::Val::I64(b)) => a == b,
+            (wasm_spec_interpreter::Value::F32(a), wasmtime::Val::F32(b)) => {
+                f32_equal(*a as u32, *b)
+            }
+            (wasm_spec_interpreter::Value::F64(a), wasmtime::Val::F64(b)) => {
+                f64_equal(*a as u64, *b)
+            }
+            (_, _) => unreachable!("fuzzing non-scalar value types is still TODO"),
+        }
+    }
+
+    match (&spec_vals, &wasmtime_vals) {
+        // Compare the returned values, failing if they do not match.
+        (Ok(spec_vals), Ok(wasmtime_vals)) => {
+            let all_match = spec_vals
+                .iter()
+                .zip(wasmtime_vals)
+                .all(|(s, w)| matches(s, w));
+            if !all_match {
+                show_wat(wasm);
+                panic!(
+                    "Values do not match: spec returned {:?}; wasmtime returned {:?}",
+                    spec_vals, wasmtime_vals
+                );
+            }
+        }
+        // If both sides fail, skip this fuzz execution.
+        (Err(spec_error), Err(wasmtime_error)) => {
+            // The `None` value returned here indicates that both sides
+            // failed--if we see too many of these we might be failing too often
+            // to check instruction semantics. At some point it would be
+            // beneficial to compare the error messages from both sides (TODO).
+            // It would also be good to keep track of statistics about the
+            // ratios of the kinds of errors the fuzzer sees (TODO).
+            warn!(
+                "Both sides failed: spec returned '{}'; wasmtime returned {:?}",
+                spec_error, wasmtime_error
+            );
+            return None;
+        }
+        // If only one side fails, fail the fuzz the test.
+        _ => {
+            show_wat(wasm);
+            panic!(
+                "Only one side failed: spec returned {:?}; wasmtime returned {:?}",
+                &spec_vals, &wasmtime_vals
+            );
+        }
+    }
+
+    // TODO Compare memory contents.
```

Comment:
I think it might be worthwhile to create a tracking issue for "Wasm semantics fully fuzzed against spec interpreter", with at least (i) this issue, i.e. compare memory contents, and also compare globals; and (ii) other Wasm features we want to enable in fuzzing, like SIMD, with the appropriate interpreter updates.

Reply:
https://github.com/bytecodealliance/wasmtime/issues/3175